platforms/core-execution/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/DefaultFileWatcherProbeRegistry.java

```diff
@@ -185,8 +203,12 @@ public File getProbeFile() {
             return probeFile;
         }
 
-        public File getWatchableHierarchy() {
-            return watchableHierarchy;
+        public Set<File> getWatchableHierarchies() {
+            return watchableHierarchies;
+        }
+
+        public void addWatchableHierarchy(File watchableHierarchy) {
+            watchableHierarchies.add(watchableHierarchy);
```

Comment:
❓ It would be great if we could do this without a mutable `HashSet`. I think we should be able to collect all the file hierarchies to probe at one point, and associate them with file systems, and then create the probes in one go, allowing us to record the hierarchies as an immutable collection.